File: Auth/Auth.App/JwtTokenService.cs
using Auth.App.Env;
using Auth.App.Exceptions;
using Auth.App.Interface.RefreshToken;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Linq;
using System.Security.Claims;
using System.Security.Cryptography;
using Users.App;
using Users.App.Interface;
using Users.Domain;

namespace Auth.App
{
    public class JwtTokenService(IOptions<JwtConfig> jwtConfig, RefreshTokenService rtService, 
        LoginService loginService, IUserRepo users, ILogger<JwtTokenService> logger)
    {
        private readonly JwtConfig _jwtConfig = jwtConfig.Value;

        public async Task<TokenPair?> RefreshTokensAsync(string refreshToken)
        {
            // 1) Lookup the record
            var record = await rtService.RefreshAsync(refreshToken);
            
            if(record == null)
            {
                return null;
            }

            // 3) Issue new JWT + RT
            var user = await users.FindByIdAsync(record.UserId);
            
            if (user == null)
            {
                logger.LogWarning("RefreshTokensAsync: Could not find {UserId} via refresh token", record.UserId);
                return null;
            }

            return await IssueTokens(user);
        }

        public async Task RevokeRefreshTokenAsync(string refreshToken, bool revokeAll)
        {
           await rtService.RevokeAsync(refreshToken, revokeAll);
        }

        public async Task<TokenPair?> IssueTokenAsync(string username, string password)
        {
            var user = await loginService.LoginAsync(new LoginRequest(username, password, true, _jwtConfig.Issuer));
            if (user == null)
            {
                return null;
            }

            return await IssueTokens(user);
        }

        public async Task<TokenPair?> IssueTokensForFederatedUser(string externalId)
        {
            var user = await loginService.LoginFederatedAsync(externalId, true);
            if(user == null)
            {
                return null;
            }

            return await IssueTokens(user);
        }

        public async Task<TokenPair> IssueTokensForUserAsync(User user)
        {
            return await IssueTokens(user);
        }
       

        private async Task<TokenPair> IssueTokens(User user)
        {
            var jwt = GenerateJwtToken(user);
            var rToken = await rtService.GenerateAsync(user.Id);
            return new TokenPair(jwt, rToken, _jwtConfig.ExpiresInSeconds, user);
        }

        private string GenerateJwtToken(User user)
        {
            var keyBytes = Convert.FromBase64String(_jwtConfig.Key);
            var creds = new SigningCredentials(new SymmetricSecurityKey(keyBytes), SecurityAlgorithms.HmacSha256);
            var expires = DateTime.UtcNow.AddSeconds(_jwtConfig.ExpiresInSeconds);

            var claims = new List<Claim> {
                new(JwtRegisteredClaimNames.Sub, user.Id.ToString()),
                new(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString())
            };

            if(user.Name != null)
            {
                claims.Add(new(JwtRegisteredClaimNames.Name, user.Name));
            }

            if (user.Email != null)
            {
                claims.Add(new(JwtRegisteredClaimNames.Email, user.Email));
            }

            claims.AddRange((user.Roles?.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries) 
                ?? Array.Empty<string>()).Select(x => new Claim(ClaimTypes.Role, x)));

            claims.Add(new("idp", Constants.GetIdp(user.Issuer)));

            var token = new JwtSecurityToken(
                issuer: _jwtConfig.Issuer,
                audience: _jwtConfig.Audience,
                claims: claims,
                expires: expires,
                signingCredentials: creds
            );
           
            var tokenStr = new JwtSecurityTokenHandler().WriteToken(token);

            return tokenStr;
        }
    }
}

File: NodeAuth/src/services/JwtTokenService.ts
import { randomUUID } from "crypto";
import jwt from "jsonwebtoken";

import { JwtConfig } from "../config/jwt";
import { TokenPair } from "../models/TokenPair";
import { User } from "../models/User";
import { IUserRepo } from "../repos/interfaces/IUserRepo";
import { RefreshTokenService } from "./RefreshTokenService";

const roleClaimType =
  "http://schemas.microsoft.com/ws/2008/06/identity/claims/role";
const knownIdps = ["apple", "google", "microsoft"];

export class JwtTokenService {
  constructor(
    private readonly config: JwtConfig,
    private readonly refreshTokenService: RefreshTokenService,
    private readonly users: IUserRepo,
  ) {}

  async refreshTokensAsync(refreshToken: string): Promise<TokenPair | null> {
    const record = await this.refreshTokenService.refreshAsync(refreshToken);
    if (!record) {
      return null;
    }

    const user = await this.users.findByIdAsync(record.userId);
    if (!user) {
      return null;
    }

    return this.issueTokens(user);
  }

  async revokeRefreshTokenAsync(
    refreshToken: string,
    revokeAll: boolean,
  ): Promise<void> {
    await this.refreshTokenService.revokeAsync(refreshToken, revokeAll);
  }

  async issueTokenAsync(user: User): Promise<TokenPair> {
    return this.issueTokens(user);
  }

  private async issueTokens(user: User): Promise<TokenPair> {
    const accessToken = this.generateJwtToken(user);
    const refreshToken = await this.refreshTokenService.generateAsync(user.id);
    return {
      accessToken,
      refreshToken,
      expiresIn: this.config.expiresInSeconds,
      user,
    };
  }

  private generateJwtToken(user: User): string {
    const key = Buffer.from(this.config.key, "base64");
    const roles = (user.roles ?? "")
      .split(",")
      .map((role) => role.trim())
      .filter((role) => role.length > 0);
    const payload: jwt.JwtPayload = {
      sub: String(user.id),
      jti: randomUUID(),
      idp: this.getIdp(user.issuer),
    };

    if (user.name) {
      payload.name = user.name;
    }

    if (user.email) {
      payload.email = user.email;
    }

    if (roles.length > 0) {
      payload[roleClaimType] = roles;
    }

    return jwt.sign(payload, key, {
      algorithm: "HS256",
      issuer: this.config.issuer,
      audience: this.config.audience,
      expiresIn: this.config.expiresInSeconds,
    });
  }

  private getIdp(issuer: string): string {
    const normalized = issuer.toLowerCase();
    return knownIdps.find((idp) => normalized.includes(idp)) ?? issuer;
  }
}
